### ISSUE TITLE
CVE-2021-27293: Fix NewDateRegex

### ISSUE THREAD

**Issue by b-c-ds** [restsharp/RestSharp#1556]
This advisory was emailed to the maintainer. Posting here as an issue as requested.

# Doyensec Vulnerability Advisory 
* Regular Expression Denial of Service (REDoS) in RestSharp
* Affected Product: RestSharp (All released versions)
* Vendor: https://restsharp.dev
* Severity: Medium
* Vulnerability Class: Denial of Service
* Status: Open
* Author: Ben Caller ([Doyensec](https://doyensec.com))

# SUMMARY

The .NET library RestSharp uses a regular expression which is vulnerable to Regular Expression Denial of Service (REDoS) when converting strings into DateTimes. If a server responds with a malicious string, the client using RestSharp will be stuck processing it for an exceedingly long time. This allows the remote server to trigger a Denial of Service.

# TECHNICAL DESCRIPTION

The vulnerable regular expression is `NewDateRegex` in `RestSharp.Extensions.StringExtensions`:

https://github.com/restsharp/RestSharp/blob/0ed7b0a6b64ab4b9838c2c0cb76a1808facebe09/src/RestSharp/Extensions/StringExtensions.cs#L28

It is used by the `ParseJsonDate` function when deserializing JSON responses into classes with `DateTime` properties.

Due to the `(-?\d+)*` part containing nested repeats, this regular expression has catastrophic backtracking when processing a long string of digits. The behaviour occurs as long as the digits are *not* followed immediately by a closing parenthesis ')'.

An example of a REDoS payload is `new Date(12345678901234567890123456789012345`.

The space between 'new' and 'Date' is required due to pre-processing in `ParseJsonDate`:

https://github.com/restsharp/RestSharp/blob/0ed7b0a6b64ab4b9838c2c0cb76a1808facebe09/src/RestSharp/Extensions/StringExtensions.cs#L124-L126

The complexity is exponential: increasing the length of the malicious string of digits by one makes processing take about twice as long. On my laptop, 27 digits takes about 16 seconds to process and 28 digits takes about 32 seconds, so a string with 54 digits should take approximately 68 years to process.

The vulnerable regular expression was first introduced in commit https://github.com/restsharp/RestSharp/commit/373a0a3d52dede087f424949db1e02e7d06287fd

# REPRODUCTION STEPS

The REDoS can be triggered by calling `RestSharp.Extensions.StringExtensions.ParseJsonDate` directly, or by deserializing JSON responses into a class with a property of type `DateTime`.

Example C# code to see the effect of the REDoS is attached below. Changing the length of the string of zeroes will change the processing time.

```c#
using System;
using System.Globalization;
using RestSharp;
using RestSharp.Extensions;
using RestSharp.Serialization.Json;

namespace DoyensecRestSharpRedosTest
{
    class Thing
    {
        public DateTime Time { get; set; }
    }

    class Program
    {
        static void ByParseJsonDate()
        {
            var redos = "new Date(" + new String('0', 30);
            DateTime dt = StringExtensions.ParseJsonDate(redos, CultureInfo.InvariantCulture);
        }

        static void ByDeserializer()
        {
            var redosJson = @"{""Time"": ""new Date(" + new String('0', 30) + @"""}";
            var thing = (new JsonDeserializer()).Deserialize<Thing>(new RestResponse {Content = redosJson});
            Console.WriteLine(thing.Time);
        }

        static void ByRequest()
        {
            // Server should return JSON {"Time": "new Date(000000000000000000000000000000"}
            Thing thing = new RestClient("https://localhost/thing.json").Execute<Thing>(new RestRequest(Method.GET)).Data;
            Console.WriteLine(thing.Time);
        }

        static void Main(string[] args)
        {
            ByDeserializer(); // or try ByParseJsonDate() or ByRequest()
        }
    }
}
```

# REMEDIATION

Fix `NewDateRegex`. We propose simply removing the asterisk:

```c#
    static readonly Regex NewDateRegex = new Regex(@"newDate\((-?\d+)\)");
```
.

Doyensec (www.doyensec.com) is an independent security research and development company focused on vulnerability discovery and remediation. We work at the intersection of software development and offensive engineering to help companies craft secure code. 

Copyright 2021 by Doyensec LLC. All rights reserved. 

Permission is hereby granted for the redistribution of this advisory, provided that it is not altered except by reformatting it, and that due credit is given. Permission is explicitly given for insertion in vulnerability databases and similar, provided 
that due credit is given. The information in the advisory is believed to be accurate at the time of publishing based on 
currently available information, and it is provided as-is, as a free service to the community by Doyensec LLC. There are 
no warranties with regard to this information, and Doyensec LLC does not accept any liability for any direct, indirect, or 
consequential loss or damage arising from use of, or reliance on, this information.

**Comment by b-c-ds** [comment]
Assigned CVE-2021-27293

**Comment by abhijeet490** [comment]
@bcaller I can see a fix to this vulnerability, but on which version can we get this update?

**Comment by b-c-ds** [comment]
@abhijeet490 The patch has not reviewed or merged. If you want the fix anytime soon you may need to make a fork.

**Comment by neil-gok** [comment]
I see that this issue has been closed, but the vulnerability still exists in version 106.11.7.  On what version was this fix released?

**Comment by b-c-ds** [comment]
106.11.8-alpha.0.13 😆

**Comment by RajRele** [comment]
When could we expect a stable release which includes this fix?

**Comment by gdoron** [comment]
I'm honestly asking, is it the common practice to share CVE publicly **before** the fix was released and adopted?

**Comment by gavinBurtonStoreFeeder** [comment]
> I'm honestly asking, is it the common practice to share CVE publicly **before** the fix was released and adopted?

CVE's are already public knowledge. But its disappointing that its been since February without a patch.

@b-c-ds This issue should not be closed. Can you please confirm whether an official patch is planned?

**Comment by b-c-ds** [comment]
@gavinBurtonStoreFeeder I'm not a maintainer. It won't let me reopen the issue. Sorry. I only reported this issue and contributed one pull request. I don't think this project is super active so I wouldn't hold my breath waiting for a non-alpha release.
Also, if you aren't deserialising DateTimes you should be able to ignore this vulnerability.

**Comment by gavinBurtonStoreFeeder** [comment]
> @gavinBurtonStoreFeeder I'm not a maintainer. It won't let me reopen the issue. Sorry. I only reported this issue and contributed one pull request. I don't think this project is super active so I wouldn't hold my breath waiting for a non-alpha release.
> Also, if you aren't deserialising DateTimes you should be able to ignore this vulnerability.

Fair enough. Thanks for raising the GH issue. There is a fix underway but I feel like restsharp is abandonware now.

**Comment by alexeyzimarev** [comment]
> its been since February without a patch.

The PR was included to 106.12.0 release

>  restsharp is abandonware now.

It is a typical ".NET community project", isn't it? 99M downloads, zero sponsors, little contributions from non-maintainers, but lots of complaints (like the comment just above).

Big kudos to @b-c-ds for reporting and fixing the issue, it's now released, the issue is closed.

**Comment by gavinBurtonStoreFeeder** [comment]
> > its been since February without a patch.
> 
> The PR was included to 106.12.0 release
> 
> > restsharp is abandonware now.
> 
> It is a typical ".NET community project", isn't it? 99M downloads, zero sponsors, little contributions from non-maintainers, but lots of complaints (like the comment just above).
> 
> Big kudos to @b-c-ds for reporting and fixing the issue, it's now released, the issue is closed.

You quoted me out of context by starting halfway through the sentence. And I made that comment in july, 5 months after the PR was created.
